### PR TITLE
[lgwebos] Avoid thing updates when the thing handler is already disposed

### PR DESCRIPTION
**Pull request by lolodomo**
Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java [hunk 1]
```diff
@@ -332,6 +333,10 @@ private void sendMessage(JsonObject json) {
     @OnWebSocketMessage
     public void onMessage(String message) {
         logger.trace("Message [in]: {}", message);
+        if (this.listener == null) {
+            logger.debug("Message is ignored because the thing handler is already disposed.");
+            return;
+        }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: Is there a more direct way of determining if the handler is disposed? Or more importantly, shouldn't this class be disposed along with the handler in the first place?

Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: In fact, this is the more direct way I found.
Here is the code run when the thing handler is disposed:
```
        LGWebOSTVSocket s = socket;
        if (s != null) {
            s.setListener(null);
            scheduler.execute(() -> s.disconnect()); // dispose should be none-blocking
        }
```
As you can see, this is done asynchronously. It can happen that messages are received and handled bedore `disconnect()` is run.

setListener is called with the thing handler in thing handler initialize().
setListener is called with null in thing handler dispose().

Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: The variable is named `listener` but in fact listener = thingHandler

Comment (cpmeister) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: I'm not sure that `s.disconnect()` should be called asynchronously as part of disposing. Any essential cleanup should be done immediately because `initialize()` might be called immediately afterward and any socket resources getting initialized might collide with the socket resources that are getting disposed.

But I'm not sure if such a change is within the scope of this PR...

Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: @sprehn : can you comment about the asynchronous disconnect please ? The comment looks clear, is s.disconnect() really blocking ? Until what ?

@cpmeister : yes, the scope of my PR was not to change that but rather avoid having the reconnect job continuing its scenario while the thing handler is already disposed.

**Comment by sprehn** [comment]
we indeed should call s.disconnect synchronously 
all it does is call a close in the websocket, which is asynchronous itself according to docs.
but subsequently it will set the state to disconnected.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java [hunk 2]
```diff
@@ -332,6 +333,10 @@ private void sendMessage(JsonObject json) {
     @OnWebSocketMessage
     public void onMessage(String message) {
         logger.trace("Message [in]: {}", message);
+        if (this.listener == null) {
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:336]: Don’t like this change 
this class should be agnostic to OH types like thing handler and lifecycle

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java (hunk 1 above)
Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: yes we indeed can make this call synchronous 
s.disconnect enqueues a websocket close, which is asynchronous 
it synchronously sets the state to disconnected

Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: the lgweostvsocket instance gets disposed 
a new instance is created during initialize

can you remind me what the issues is that needs to be solved?

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:336]: This class is already linked to OH as you are updating the thing properties and the thing configuration when handling received messages (during the registering process). Of course, we have to avoid that when the thing handler is already disposed.
But we could moved this test before updating thing properties/configuration.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java:339]: You can read again last @cpmeister message, he explained the problem. It occurs when the thing handler is quickly disposed and initialized again while the first registering process is still running.

**Comment by lolodomo** [comment]
Ok, I just updated the PR, the problem is now handled by the thing handler.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java (hunk 1 above)
Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java]: but we create a new socket each time we initialize. so states should be clean. 

what is the issue for end users?

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java (hunk 2 above)
Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSTVSocket.java]: therefore there is an interface which thinghandler implements

the behavior should not change whether or not somebody is whatching. reminds me of schrödinger’s cat

**Comment by sprehn** [comment]
still trying to understand the issue. 

OH initializes the handler
the reconnect job starts
once the socket connection can be established binding sends "hello" message to TV
tv responds with hello response
binding calls config.storeProperties(map);
binding sends "register" message to TV
tv responds with register response 
binding calls config.storeKey(key);

I agree that we could optimize both storeProperties and storeKey to only actually update if values changed. It may already solve the problem, as these values are only set once at the beginning.

storeKey today is already designed not to go through re-initialization. Is storeProperties the problem?

Or is the problem, that the the connection sequence continues running, while the binding disposes, calls disconnect and LGWebOSTVSocket is in state DISCONNECTED. If so, what we could do is to stop the sequence by putting entry conditions into the connection sequence of LGWebOSTVSocket::onMessage

// add additional state CONNECTING
```
switch (response.getType()) {
....
            case "hello":
               if (state != CONNECTING) {
                    logger.debug("Skipping resonse {}, not in CONNECTING state, state was {}", message, state);
                    break;
                }
....
                break;
            case "registered":
               if (state != REGISTERING) {
                    logger.debug("Skipping resonse {}, not in REGISTERING state, state was {}", message, state);
                    break;
                }
  ....
              break;
        }
```

**Comment by lolodomo** [comment]
> Or is the problem, that the the connection sequence continues running, while the binding disposes

Yes, you got it.

Your option looks good to me, it is more or less the same as my previous test "listener == null" but with a really cleaner solution.
If a new state CONNECTING is introducing, we just have to take care of it in the thing handler.
I will change the PR to use this way to implement the fix.

**Comment by lolodomo** [comment]
@sprehn : please review this new version.

**Comment by lolodomo** [comment]
@sprehn : just one question/detail: we stop reconnection job and start keep alive job when the state is changed to REGISTERING. Should we rather do it now when the state is changed to CONNECTING ?

**Comment by lolodomo** [comment]
@cpmeister : can you start a new review please ? The PR was largely changed since your first approval.

**Comment by sprehn** [comment]
> @sprehn : just one question/detail: we stop reconnection job and start keep alive job when the state is changed to REGISTERING. Should we rather do it now when the state is changed to CONNECTING ?

Good point. 

It is better to stopReconnectJob() in CONNECTING.
but startKeepAliveJob should actually move to REGISTERED.

**Comment by lolodomo** [comment]
Change with jobs done

**Comment by cpmeister** [comment]
Manually approved sign-off
